File: iafactory-gateway-python/providers/groq.py
# -*- coding: utf-8 -*-
"""
Provider Groq
"""
import httpx
import os

GROQ_API_KEY = os.getenv("GROQ_API_KEY", "")
GROQ_BASE_URL = "https://api.groq.com/openai/v1"

class GroqProvider:
    def __init__(self):
        self.api_key = GROQ_API_KEY
        self.base_url = GROQ_BASE_URL
    
    async def chat_completion(self, model: str, messages: list, temperature: float = 0.7, max_tokens: int = 1000, stream: bool = False):
        """Appel chat completion Groq"""
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }
        
        payload = {
            "model": model,
            "messages": messages,
            "temperature": temperature,
            "max_tokens": max_tokens,
            "stream": stream
        }
        
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(
                f"{self.base_url}/chat/completions",
                headers=headers,
                json=payload
            )
            response.raise_for_status()
            return response.json()

groq_provider = GroqProvider()

File: iafactory-gateway/src/providers/groq.ts
import Groq from 'groq-sdk';
import { appConfig } from '../config.js';
import type {
  BaseProvider,
  ChatCompletionRequest,
  ChatCompletionResponse,
  StreamChunk,
} from './types.js';

export class GroqProvider implements BaseProvider {
  name = 'groq';
  private client: Groq;

  constructor() {
    if (!appConfig.GROQ_API_KEY) {
      throw new Error('GROQ_API_KEY not configured');
    }

    // Ensure baseURL does not double-concatenate /openai/v1
    // If GROQ_BASE_URL ends with /openai/v1, use as is; otherwise, append /openai/v1
    let baseURL = appConfig.GROQ_BASE_URL;
    if (baseURL && !baseURL.endsWith('/openai/v1')) {
      baseURL = baseURL.replace(/\/+$/, '') + '/openai/v1';
    }

    this.client = new Groq({
      apiKey: appConfig.GROQ_API_KEY,
      baseURL,
    });
  }

  async complete(request: ChatCompletionRequest): Promise<ChatCompletionResponse> {
    const response = await this.client.chat.completions.create({
      model: request.model,
      messages: request.messages,
      temperature: request.temperature,
      max_tokens: request.max_tokens,
      top_p: request.top_p,
      frequency_penalty: request.frequency_penalty,
      presence_penalty: request.presence_penalty,
      stream: false,
    });

    return {
      id: response.id,
      object: 'chat.completion',
      created: response.created,
      model: response.model,
      choices: response.choices.map((choice) => ({
        index: choice.index,
        message: {
          role: choice.message.role as 'assistant',
          content: choice.message.content || '',
        },
        finish_reason: choice.finish_reason || 'stop',
      })),
      usage: response.usage
        ? {
            prompt_tokens: response.usage.prompt_tokens,
            completion_tokens: response.usage.completion_tokens,
            total_tokens: response.usage.total_tokens,
          }
        : undefined,
    };
  }

  async *streamComplete(request: ChatCompletionRequest): AsyncGenerator<StreamChunk> {
    const stream = await this.client.chat.completions.create({
      model: request.model,
      messages: request.messages,
      temperature: request.temperature,
      max_tokens: request.max_tokens,
      top_p: request.top_p,
      stream: true,
    });

    for await (const chunk of stream) {
      const choice = chunk.choices[0];
      
      yield {
        id: chunk.id,
        object: 'chat.completion.chunk',
        created: chunk.created,
        model: chunk.model,
        delta: {
          role: choice.delta.role,
          content: choice.delta.content || '',
        },
        finish_reason: choice.finish_reason || undefined,
        usage: chunk.x_groq?.usage
          ? {
              prompt_tokens: chunk.x_groq.usage.prompt_tokens,
              completion_tokens: chunk.x_groq.usage.completion_tokens,
              total_tokens: chunk.x_groq.usage.total_tokens,
            }
          : undefined,
      };
    }
  }
}
